### ISSUE TITLE
Recreating broken  DB takes forever

### ISSUE THREAD

**Issue by mdsmdsmds** [duplicati/duplicati#2889]
Duplicati - 2.0.2.12_canary_2017-10-20
Windows Server 2012 R2

- no large fileset (I've 18GB of disk used on local disk repository)
- Recreation starts and this is log:

Nov 11, 2017 12:53 AM: Message
      Processing all of the 223 volumes for blocklists
Nov 11, 2017 12:48 AM: Message
      Filelists restored, downloading 223 index files
Nov 11, 2017 12:22 AM: Message
      Rebuild database started, downloading 20 filelists

- after that, all stay there forever. It keeps reading and writing on temp files but never end.
- the DB seems almost complete (RFOJWUPAWF.sqlite is now 241MB) [this is its previous tipical size]
- but non more progressing. It stay there forever
- After 12 hours recreating is still there. Look at screenshot...
![screenshot - 11_11_2017 13_23_46](https://user-images.githubusercontent.com/22540148/32689577-c1e31d50-c6e7-11e7-8d60-5aae19ed8d02.png)

This seems very similar at https://github.com/duplicati/duplicati/issues/1328 opened by @Simpelton

**Comment by mdsmdsmds** [comment]
and more... the process seems freezed. I've to kill it by task manager because the stop button is unresponsive.

**Comment by kenkendk** [comment]
Yes, I am aware of the performance issues with recreating the database. I do not have a quick fix, but I will prioritize making the recreate faster.

**Comment by mdsmdsmds** [comment]
Hi, @kenkendk. Could this be related at https://github.com/duplicati/duplicati/issues/2848#issuecomment-345253288 ?
This because I noticed that, improving random disk access (forcing with disk-cache-SW), recreation is much more rapid and successfull.

**Comment by kenkendk** [comment]
That would at least work with the IO problems, but I am not sure it is enough.

**Comment by mdsmdsmds** [comment]
Duplicati opens DB with "non-cache Flag".
Colud you instead open it in "cache-mode"?

![screenshot - 15_12_2017 15_39_54](https://user-images.githubusercontent.com/22540148/34046945-0429c4f2-e1af-11e7-8945-b90c0988a482.png)

**Comment by ConradHughes** [comment]
I suspect you probably understand the reasons for slow database recreation quite thoroughly, but here's a datapoint: a recent database recreate for a 1.4TB backup took me 14 days, during which there were several worrying multi-day periods of no movement on the 610-pixel-wide progress meter (yes, I started counting pixels).  The machine largely seemed to be I/O constrained by its physical hard drives: CPU was rarely very high but my NAS, which was running Duplicati, reported volume I/O to be maxed out every time I looked.

**Comment by sirius-c** [comment]
Another (incomplete) datapoint: I had to recreate the database of a 190 GByte local backup with 10 versions (minimal changes between versions). Database recreation went speedy initially, but then only crept forward (from 94% to 95% completion within 24 hours). I eventually stopped the recreation after more than two days, deleted the backup, and created a completely new one - this took 50 minutes.

This is on a recent quad-core computer, SSD for the operating system, backup destination on separate local harddisk. Duplicati - 2.0.4.5_beta_2018-11-28, Windows 10. 

Fixing this incredible slow database recreation would improve the usability and trust in Duplicati substantially to my opinion.

**Comment by torbennehmer** [comment]
Might this be a duplicate of #3668 (or vice versa)?

**Comment by t0mcat1337** [comment]
Me also can confirm this issue... cause of an encryption - virus (almost each and every file was encrypted, so the Duplicati Database too) we had to completely Re-Install a Windows Server.
Now we installed Duplicati 2.0.4.5_beta_2018-11-28 and choose "Restore from Files", so that Duplicati will recreate a temporary Database from the Backup files during the restore. 
Creating this temporary database lasts 2 days now! First, the progress bar moves relativly fast and then suddenly seems to stuck at ~85-90% . But then we noticed, that it moves extreeeeemly slow for the last ~15-10%.

The Backup Files are 97GB in size and we copied them from the FTP Site to the servers local Disk to make sure that FTP can't be a bottleneck.

On another Machine we tried an older Version of Duplicati (2.0.1.61) but there the problem is the same...

Is there anything that could be done to speed things up? Recreating a Backup - Meta - Data - Database may never last days(!) IMHO

**Comment by joel0** [comment]
Not only is this a slow operation, but it writes (and deletes) quite a bit to the temp folder.  I have a 600 GB backup on with 15 versions that I'm repairing.  Just in the span of one day, I've noticed that my SSD's total bytes written has increased by over 4 TB (if my memory serves correctly)!  I would move it to my HDD, but since the data access looks very random, I assume it would make the process insanely slow.  With only 1 pixel of progress in the span of about 4+ hours, it looks like it would be quicker (and safer for my SSD) to delete the whole backup and spend a week re-uploading everything.

Until the repair process is optimised, I think it might be worthwhile putting a warning about potential excessive SSD wear before starting the repair.  Extrapolating my numbers, the repair operation could have written enough data to void my SSD's warranty.

**Comment by orappold** [comment]
I'm also very unhappy about this issue. This way I do not have any trust in Duplicati where it already consumes so much of my time to get my backups working. Very slow recreating db times and also verifying backup files takes ages. 

@joel0: Thanks for mentioning this, I would have not thought about it but of course it has an major impact on my SSD as well. 

I tried the recent beta and the latest canary version - both do not work well enough for me. 
I will uninstall duplicati and look for an alternative.

**Comment by roterBulle** [comment]
Maybe you can try stopping the service and opening the database with a sqlite tool (like DB Browser for SQL Lite and running the ANALYZE command. After the command has finished rerun your recreate job.
Atleast for me this helped fixing the issue of slow SQL queries.

**Comment by joel0** [comment]
@roterBulle In my case, there was a lot of bandwidth usage (downloading many dblocks totaling hundreds of GB) and disk utilization (writing over 4 TB within part of a day), so I'd guess that slow sqlite queries wasn't the primary problem I was facing.

Since I resolved my problem (by pointing the service to the correct sqlite database instead of the new, empty one it started creating), I can't easily test running the analyze command.

**Comment by dyc3** [comment]
Looking at my logs, the database inserts take 2-3 minutes, compared to the downloads which take about 3 seconds each. I think we can pretty safely say that the bottleneck is the database operations.

**Comment by dyc3** [comment]
I've noticed that these insert queries are huge af. Here's an example:

```sql
INSERT INTO "BlocksetEntry" ("BlocksetID", "Index", "BlockID") SELECT DISTINCT "H"."BlocksetID", "H"."Index", "H"."BlockID" FROM (SELECT "E"."BlocksetID" AS "BlocksetID", "D"."FullIndex" AS "Index", "F"."ID" AS "BlockID" FROM ( SELECT "E"."BlocksetID", "F"."Index" + ("E"."BlocklistIndex" * 3200) AS "FullIndex", "F"."BlockHash", MIN(102400, "E"."Length" - (("F"."Index" + ("E"."BlocklistIndex" * 3200)) * 102400)) AS "BlockSize", "E"."Hash", "E"."BlocklistSize", "E"."BlocklistHash" FROM ( SELECT * FROM ( SELECT "A"."BlocksetID", "A"."Index" AS "BlocklistIndex", MIN(3200 * 32, ((("B"."Length" + 102400 - 1) / 102400) - ("A"."Index" * (3200))) * 32) AS "BlocklistSize", "A"."Hash" AS "BlocklistHash", "B"."Length" FROM "BlocklistHash" A, "Blockset" B WHERE "B"."ID" = "A"."BlocksetID" ) C, "Block" D WHERE "C"."BlocklistHash" = "D"."Hash" AND "C"."BlocklistSize" = "D"."Size" ) E, "TempBlocklist-769073E2EB268C40801CCFDEFB3CED78" F WHERE "F"."BlocklistHash" = "E"."Hash" ORDER BY "E"."BlocksetID", "FullIndex" ) D, "BlocklistHash" E, "Block" F, "Block" G WHERE "D"."BlocksetID" = "E"."BlocksetID" AND "D"."BlocklistHash" = "E"."Hash" AND "D"."BlocklistSize" = "G"."Size" AND "D"."BlocklistHash" = "G"."Hash" AND "D"."Blockhash" = "F"."Hash" AND "D"."BlockSize" = "F"."Size" UNION SELECT "Blockset"."ID" AS "BlocksetID", 0 AS "Index", "Block"."ID" AS "BlockID" FROM "Blockset", "Block", "TempSmalllist-532CE48AB7955A43A98CA6605AD57B0B" S WHERE "Blockset"."Fullhash" = "S"."FileHash" AND "S"."BlockHash" = "Block"."Hash" AND "S"."BlockSize" = "Block"."Size" AND "Blockset"."Length" = "S"."BlockSize" AND "Blockset"."Length" <= 102400 ) H WHERE ("H"."BlocksetID" || ':' || "H"."Index") NOT IN (SELECT ("ExistingBlocksetEntries"."BlocksetID" || ':' || "ExistingBlocksetEntries"."Index") FROM "BlocksetEntry" "ExistingBlocksetEntries" )
```

Honestly, I'm not 100% sure what its doing, but I think it's doing a bunch of selects to compute some stuff to insert it into the `BlocksetEntry` table. Normally, I'd say to just parallelize it, but **sqlite doesn't allow for multiple concurrent writers**. However, we can safely have multiple readers.

Here's some ideas:
We could batch the inserts, minimizing writes, and move some of this extra processing out of SQL and into C# land. Then we could multithread some of this processing, and easily flush these inserts all in one go. I'm not too familiar with the internals, so I'm not entirely sure how feasible that is though.

**Comment by duplicatibot** [comment]
This issue has been mentioned on **Duplicati**. There might be relevant details there:

https://forum.duplicati.com/t/need-help-resuming-local-database-repair/13050/1

**Comment by duplicatibot** [comment]
This issue has been mentioned on **Duplicati**. There might be relevant details there:

https://forum.duplicati.com/t/need-help-resuming-local-database-repair/13050/3

**Comment by samuel-w** [comment]
It's possible to have multiple writers, requires write-ahead logging. https://sqlite.org/src/doc/begin-concurrent/doc/begin_concurrent.md

I tried enabling write-ahead logging alone but it didn't really change my recreate times.
